packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -351,166 +375,213 @@ export interface EmbeddedChunk<ChunkMetadata extends Jsonifiable = Jsonifiable>
+   * Note: Corpus loading may be quite time consuming. It's best to ensure loading is done prior
+   * to exposing this corpus to your users.
+   *
+   * Note: This method is idempotent. Calling it multiple times will not result in multiple loads.
+   */
+  async load(): Promise<CorpusStats> {
+    if (!this.loadingPromise) {
+      this.stats.loadingState = CorpusLoadingState.IN_PROGRESS;
+      this.loadingPromise = loadCorpus(this.loader, this.vectorizer, this.chunkConsumer);
     }
-    this.loadingState = Corpus.LoadingState.IN_PROGRESS;
     try {
-      await this.load();
-      this.loadingState = Corpus.LoadingState.COMPLETED;
+      this.stats = await this.loadingPromise;
     } catch (e) {
-      this.loadingState = Corpus.LoadingState.FAILED;
+      this.stats.loadingState = CorpusLoadingState.FAILED;
+      if (e instanceof Error) {
+        this.stats.loadingError = e;
+      }
       throw e;
     }
-    return this.getStats();
+    return this.stats;
   }
 
-  private async load(): Promise<void> {
-    this.activePartitionsToToken.set(null, null);
-    while (this.activePartitionsToToken.size > 0) {
-      const [partition, pageToken] = this.activePartitionsToToken.entries().next().value;
-      this.activePartitionsToToken.delete(partition);
+  /** Reveals statistics about this corpus, including its LoadingState. */
+  getStats(): CorpusStats {
+    return this.stats;
+  }
 
-      const response = await this.loader({ partition, pageToken } as CorpusLoadRequest);
-      if (!response.page && !response.partitions) {
-        throw Error('Loader responses must include a page, new partitions, or both.');
-      }
-      for (const newPartition of response.partitions ?? []) {
-        if (!this.completedPartitions.has(newPartition.name) && !this.activePartitionsToToken.has(newPartition.name)) {
-          this.activePartitionsToToken.set(newPartition.name, newPartition.firstPageToken ?? null);
-        }
+  protected checkReady(): void {
+    if (this.stats.loadingState !== CorpusLoadingState.COMPLETED) {
+      throw new CorpusNotReadyError(this.stats.loadingState);
+    }
+  }
+
+  abstract search(
+    query: string,
+    params?: { limit?: number; score_threshold?: number }
+  ): Promise<ScoredChunk<ChunkMetadata>[]>;
+}
+
+async function loadCorpus<
+  DocumentMetadata extends Jsonifiable = Jsonifiable,
+  ChunkMetadata extends Jsonifiable = Jsonifiable
+>(
+  loader: Loader<DocumentMetadata>,
+  vectorize: Vectorizer<DocumentMetadata, ChunkMetadata>,
+  chunkConsumer: (chunks: EmbeddedChunk<ChunkMetadata>[]) => Promise<void>,
+  maxRequests: number = 50
+): Promise<CorpusStats> {
+  const activePartitionsToToken = new Map<CorpusPartition['name'] | null, string | null>();
+  const completedPartitions = new Set<CorpusPartition['name']>();
+  let numDocuments = 0;
+  let numChunks = 0;
+  let numRequests = 0;
+
+  activePartitionsToToken.set(null, null);
+  while (activePartitionsToToken.size > 0 && numRequests++ < maxRequests) {
```

Comment:
Is `maxRequests` a protection against the "crawl all of wikipedia" case?

Reply:
If a crawl stops due to hitting max requests, as opposed to exhausting all the partitions, is that visible to the dev?

Reply:
That and to prevent an infinite loop. I suppose it's not our resources though, so I upped the default to 5000 requests.

Devs could infer max requests were hit by viewing the corpus stats.  Non-zero active partitions with COMPLETED state implies max requests limit was reached.